index.js

```diff
@@ -0,0 +1,162 @@
+
+class Scanner {
+  constructor (text) {
+    this.pos = 0
+    this.text = text
+  }
+
+  eof () {
+    return this.pos >= this.text.length
+  }
+
+  next () {
+    const token = this.peek()
+    this.pos += token.length
+    return token
+  }
+
+  peek () {
+    let token = this.text.charAt(this.pos)
+    // Consume <CR>? <LF>
+    if (token === '\r' && this.text.charAt(this.pos + 1) === '\n') {
+      token += '\n'
+    }
+    return token
+  }
+
+  position () {
+    return this.pos
+  }
+
+  rewind (pos) {
+    this.pos = pos
+  }
+
+  consumeWhitespace () {
+    while (isWhitespace(this.peek())) {
+      this.pos++
+    }
+  }
+}
+
+function isParens (token) {
+  return token === '(' || token === ')'
+}
+
+/*
+ * <ZWNBSP>       ::= "U+FEFF"
+ * <TAB>          ::= "U+0009"
+ * <VT>           ::= "U+000B"
+ * <FF>           ::= "U+000C"
+ * <SP>           ::= "U+0020"
+ * <NBSP>         ::= "U+00A0"
+*/
+function isWhitespace (token) {
+  return token === '\ufeff' || token === '\u0009' || token === '\u000b' || token === '\u000c' || token === '\u0020' || token === '\u00a0'
```

Comment:
We could make these tokens a bit more readable if we use something like `codes.tab`, [like micromark](https://github.com/micromark/micromark/blob/main/lib/character/codes.mjs)